google/cloud/storage/internal/curl_client.cc

```diff
@@ -788,6 +834,188 @@ void CurlClient::LockShared() { mu_.lock(); }
+  // QuotaUser cannot be set, checked by the caller.
+
+  // TODO(#937) - use client options to configure buffer size.
+  std::unique_ptr<CurlReadStreambuf> buf(new CurlReadStreambuf(
+      builder.BuildDownloadRequest(std::string{}), 128 * 1024));
+  return std::make_pair(Status(),
+                        std::unique_ptr<ObjectReadStreambuf>(std::move(buf)));
+}
+
+std::pair<Status, std::unique_ptr<ObjectWriteStreambuf>>
+CurlClient::WriteObjectXml(InsertObjectStreamingRequest const& request) {
+  CurlRequestBuilder builder(xml_upload_endpoint_ + "/" +
+                                 request.bucket_name() + "/" +
+                                 request.object_name(),
+                             xml_upload_factory_);
+  builder.SetMethod("PUT")
+      .SetDebugLogging(options_.enable_http_tracing())
+      .SetCurlShare(share_.get())
+      .AddHeader(options_.credentials()->AuthorizationHeader())
+      .AddHeader("Host: storage.googleapis.com");
+
+  //
+  // Apply the options from InsertObjectMediaRequest that are set, translating
+  // to the XML format for them.
+  //
+  builder.AddOption(request.GetOption<ContentEncoding>());
+  // Set the content type of a sensible value, the application can override this
+  // in the options for the request.
+  if (not request.HasOption<ContentType>()) {
+    builder.AddHeader("content-type: application/octet-stream");
+  } else {
+    builder.AddOption(request.GetOption<ContentType>());
+  }
+  builder.AddOption(request.GetOption<EncryptionKey>());
+  if (request.HasOption<IfGenerationMatch>()) {
+    builder.AddHeader("x-goog-if-generation-match: " +
+        std::to_string(request.GetOption<IfGenerationMatch>().value()));
+  }
+  // IfGenerationNotMatch cannot be set, checked by the caller.
+  if (request.HasOption<IfMetagenerationMatch>()) {
+    builder.AddHeader("x-goog-if-meta-generation-match: " +
+        std::to_string(request.GetOption<IfMetagenerationMatch>().value()));
+  }
+  // IfMetagenerationNotMatch cannot be set, checked by the caller.
+  if (request.HasOption<KmsKeyName>()) {
+    builder.AddHeader("x-goog-encryption-kms-key-name: " +
+        request.GetOption<KmsKeyName>().value());
+  }
+  if (request.HasOption<PredefinedAcl>()) {
+    builder.AddHeader(
+        "x-goog-acl: " +
+            XmlMapPredefinedAcl(request.GetOption<PredefinedAcl>().value()));
+  }
+  builder.AddOption(request.GetOption<UserProject>());
+
+  //
+  // Apply the options from GenericRequestBase<> that are set, translating
+  // to the XML format for them.
+  //
+  // Fields cannot be set, checked by the caller.
+  builder.AddOption(request.GetOption<IfMatchEtag>());
+  builder.AddOption(request.GetOption<IfNoneMatchEtag>());
+  // QuotaUser cannot be set, checked by the caller.
+
+  // TODO(#937) - use client options to configure buffer size.
+  std::unique_ptr<internal::CurlStreambuf> buf(
+      new internal::CurlStreambuf(builder.BuildUpload(), 128 * 1024));
```

Comment:
Could you define this buffer size in a constant somewhere and leave a comment explaining why this value was chosen?

Reply:
Fixed.